dish_grpc_mqtt.py

```diff
@@ -102,7 +136,9 @@ def cb_add_item(key, val, category):
             if not "dish_{0}".format(category) in data:
                 data["dish_{0}".format(category)] = {}
 
-            data["dish_{0}".format(category)].update({key: val})
+            # Skip NaN values that occur on startup because they can upset Javascript JSON parsers
+            if not ((type(val) == float) and math.isnan(val)):
```

Comment:
This looks fine, but I'm wondering if this is also a problem for the other output scripts, in which case it might be better handled at a lower layer. What conditions lead to this?

BTW: In my earlier testing, I also had problems when the lower layer returned None values, which the json module translated into "null". But that may have been poor JSON parsing on the server side.

Reply:
The issue is that builtin JSON in Python admits that it is not standards compliant but takes the attitude of 'yeah, it'll be fine the other end should work it out' but in Javascript the JSON library is strict on standards compliance and fails to parse the entire JSON entity.

The simplejson library takes the approach of turning everything that isn't able to be encoded into null.  This is also what the Javascript JSON library does.

The issue only occurs after dishy has just booted and one of the obstruction related values is reported as NaN for the first few minutes.

I'm easy either way - drop the NaN values or convert them to null.

Reply:
Dropping NaNs is fine, as you have it. I only used the --json option to test, and I suspect the (Perl) server I'm using to test against is just not doing a very good job at parsing, as it also didn't like the exponential notation numbers.